### PR TITLE
Limit lightning strike effect distance

### LINKED ISSUE #1436: Excessive thunderstorms/lightning strikes
**Is your feature request related to a problem? Please describe.**
Vanilla Minecraft bases # of lightning strikes on # of loaded chunks: `For each loaded chunk, every tick there is a 1⁄100,000 chance of an attempted lightning strike during a thunderstorm.` When multiple players online, thunder becomes very common and is occasionally nearly constant.

**Describe the solution you'd like**
Configuration option to set a range for lightning strikes. Outside of this range the sound and flash are cancelled. 

**Describe alternatives you've considered**
Disabling thunderstorms completely isn't ideal because it would remove features from the game - such as charged creepers (and apparently some 1.13 features too).

**Additional context**
Default range for the sound/flash effect is `160,000 blocks`. Might be a way to change this number, since it's already in the code somewhere?

### PR DESCRIPTION
**Pull request by Trigary**
Resolves #1436

### CODE REVIEW

File: Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
++        if (sqrMaxLightningSoundDistance < 32) { // Vanilla value: Bolt impact sound effect with 2.0 volume
++            sqrMaxLightningSoundDistance = Double.MAX_VALUE;
++        } else {
++            sqrMaxLightningSoundDistance *= sqrMaxLightningSoundDistance;
++        }
++    }
++    
++    public double maxLightningFlashDistance;
++    private void maxLightningFlashDistance() {
++        maxLightningFlashDistance = getInt("lightning-strike-distance-limit.flash", -1);
```
Comment (hugmanrique) [inline Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch]: Wouldn't it be better to default to `512` here?

File: Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
++        sqrMaxLightningSoundDistance = getInt("lightning-strike-distance-limit.sound", -1);
++        if (sqrMaxLightningSoundDistance < 32) { // Vanilla value: Bolt impact sound effect with 2.0 volume
++            sqrMaxLightningSoundDistance = Double.MAX_VALUE;
++        } else {
++            sqrMaxLightningSoundDistance *= sqrMaxLightningSoundDistance;
++        }
++    }
++    
++    public double maxLightningFlashDistance;
++    private void maxLightningFlashDistance() {
```
Comment (aikar) [inline Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch]: merge this method with the other one, read both in same method

File: Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -229,6 +229,24 @@ public class PaperWorldConfig {
+             skeleHorseSpawnChance = 0.01D; // Vanilla value
+         }
+     }
++    
++    public double sqrMaxLightningSoundDistance;
++    private void sqrMaxLightningSoundDistance() {
++        sqrMaxLightningSoundDistance = getInt("lightning-strike-distance-limit.sound", -1);
++        if (sqrMaxLightningSoundDistance < 32) { // Vanilla value: Bolt impact sound effect with 2.0 volume
```
Comment (aikar) [inline Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch]: why 32? just make this < 0
If they set it to 0, then that should disable the sounds. -1 will be vanilla 'max'

File: Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
++            sqrMaxLightningSoundDistance = Double.MAX_VALUE;
++        } else {
++            sqrMaxLightningSoundDistance *= sqrMaxLightningSoundDistance;
++        }
++    }
++    
++    public double maxLightningFlashDistance;
++    private void maxLightningFlashDistance() {
++        maxLightningFlashDistance = getInt("lightning-strike-distance-limit.flash", -1);
++        if (maxLightningFlashDistance < 32) { // Vanilla value: see sqrMaxLightningSoundDistance
```
Comment (aikar) [inline Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch]: same -1 comment as above

File: Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
+index 1e30ba05..eed99be7 100644
+--- a/src/main/java/net/minecraft/server/WorldServer.java
++++ b/src/main/java/net/minecraft/server/WorldServer.java
+@@ -1082,7 +1082,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
+         }
+         // CraftBukkit end
+         if (super.strikeLightning(entity)) {
+-            this.server.getPlayerList().sendPacketNearby((EntityHuman) null, entity.locX, entity.locY, entity.locZ, 512.0D, this, new PacketPlayOutSpawnEntityWeather(entity)); // CraftBukkit - Use dimension, // Paper - use world instead of dimension
++            this.server.getPlayerList().sendPacketNearby((EntityHuman) null, entity.locX, entity.locY, entity.locZ, this.paperConfig.maxLightningFlashDistance, this, new PacketPlayOutSpawnEntityWeather(entity)); // CraftBukkit - Use dimension, // Paper - use world instead of dimension, // Paper - Limit lightning strike effect distance
```
Comment (aikar) [inline Spigot-Server-Patches/0372-Limit-lightning-strike-effect-distance.patch]: we don't need a double // Paper, just add `, limit lightning distance`

**Review by aikar — DISMISSED** [review]

**Comment by Trigary** [comment]
I have made the requested changes.

**Review by zachbr — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Comment by zachbr** [comment]
Rebased on master, was going to push to your branch and merge this but GitHub has informed me that I cannot push to your repo, despite telling right above this message that I can do that.

Pulled manually with 20515a30aed143711bb4ceae48f121e63ca401ed
Thanks

**Pull request closed without merging**